Unable to run dokka with single target multiplatform project and kotlin 1.9.0

Dokka tasks for a multiplatform project with a single target fail with

```
* Exception is:
org.gradle.api.internal.tasks.TaskDependencyResolveException: Could not determine the dependencies of task ':single-target-mpp-module:dokkaHtml'
Caused by: java.lang.UnsupportedOperationException: Empty collection can't be reduced.
	at org.jetbrains.dokka.gradle.kotlin.KotlinClasspathUtilsKt.classpathOf(kotlinClasspathUtils.kt:83)
	at org.jetbrains.dokka.gradle.kotlin.KotlinSourceSetGistKt$gistOf$3.call(KotlinSourceSetGist.kt:22)
	at org.jetbrains.dokka.gradle.kotlin.KotlinSourceSetGistKt$gistOf$3.call(KotlinSourceSetGist.kt)
```

**Installation**
- Dokka version: 1.8.20
- Kotlin Gradle plugin version: 1.9.0-Beta

--- a/runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/kotlin/kotlinClasspathUtils.kt
+++ b/runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/kotlin/kotlinClasspathUtils.kt
@@ -57,8 +57,14 @@ private fun KotlinCompilation.getKotlinCompileTask(kgpVersion: KotlinGradlePlugi
 }
 
 private fun KotlinCompilation.platformDependencyFiles(): FileCollection {
+    val excludePlatformDependencyFiles = project.classpathProperty("excludePlatformDependencyFiles", default = false)
+
+    if (excludePlatformDependencyFiles) return project.files()
     return (this as? AbstractKotlinNativeCompilation)
         ?.target?.project?.configurations
         ?.findByName(@Suppress("DEPRECATION") this.defaultSourceSet.implementationMetadataConfigurationName) // KT-58640
         ?: project.files()
 }
+
+private fun Project.classpathProperty(name: String, default: Boolean): Boolean =
+    (findProperty("org.jetbrains.dokka.classpath.$name") as? String)?.toBoolean() ?: default

--- a/runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/kotlin/kotlinClasspathUtils.kt
+++ b/runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/kotlin/kotlinClasspathUtils.kt
@@ -40,7 +40,7 @@ private fun KotlinCompilation.compileClasspathOf(project: Project): FileCollecti
 
 private fun KotlinCompilation.newCompileClasspathOf(project: Project): FileCollection {
     val compilationClasspath = (compileTaskProvider.get() as? KotlinCompileTool)?.libraries ?: project.files()
-    return compilationClasspath + platformDependencyFiles()
+    return compilationClasspath + platformDependencyFiles(project)
 }
 
 private fun KotlinCompilation.oldCompileClasspathOf(project: Project): FileCollection {
@@ -48,7 +48,7 @@ private fun KotlinCompilation.oldCompileClasspathOf(project: Project): FileColle
         return this.classpathOf(project)
     }
 
-    return this.compileDependencyFiles + platformDependencyFiles() + this.classpathOf(project)
+    return this.compileDependencyFiles + platformDependencyFiles(project) + this.classpathOf(project)
 }
 
 private fun KotlinCompilation.classpathOf(project: Project): FileCollection {
@@ -76,7 +76,7 @@ private fun KotlinCompilation.getKotlinCompileTask(kgpVersion: KotlinGradlePlugi
     }
 }
 
-private fun KotlinCompilation.platformDependencyFiles(): FileCollection {
+private fun KotlinCompilation.platformDependencyFiles(project: Project): FileCollection {
     val excludePlatformDependencyFiles = project.classpathProperty("excludePlatformDependencyFiles", default = false)
 
     if (excludePlatformDependencyFiles) return project.files()

--- a/runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/kotlin/kotlinClasspathUtils.kt
+++ b/runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/kotlin/kotlinClasspathUtils.kt
@@ -28,12 +28,7 @@ private fun KotlinCompilation.compileClasspathOf(project: Project): FileCollecti
         return this.classpathOf(project)
     }
 
-    val platformDependencyFiles: FileCollection = (this as? AbstractKotlinNativeCompilation)
-        ?.target?.project?.configurations
-        ?.findByName(@Suppress("DEPRECATION") this.defaultSourceSet.implementationMetadataConfigurationName) // KT-58640
-        ?: project.files()
-
-    return this.compileDependencyFiles + platformDependencyFiles + this.classpathOf(project)
+    return this.compileDependencyFiles + platformDependencyFiles() + this.classpathOf(project)
 }
 
 private fun KotlinCompilation.classpathOf(project: Project): FileCollection {
@@ -43,7 +38,7 @@ private fun KotlinCompilation.classpathOf(project: Project): FileCollection {
     val shouldKeepBackwardsCompatibility = (kgpVersion != null && kgpVersion < KotlinGradlePluginVersion(1, 7, 0))
     return if (shouldKeepBackwardsCompatibility) {
         // removed since 1.9.0, left for compatibility with < Kotlin 1.7
-        val classpathGetter= kotlinCompile::class.members
+        val classpathGetter = kotlinCompile::class.members
             .first { it.name == "getClasspath" }
         classpathGetter.call(kotlinCompile) as FileCollection
     } else {
@@ -59,4 +54,11 @@ private fun KotlinCompilation.getKotlinCompileTask(kgpVersion: KotlinGradlePlugi
     } else {
         this.compileTaskProvider.get() as? KotlinCompile // introduced in 1.8.0
     }
+}
+
+private fun KotlinCompilation.platformDependencyFiles(): FileCollection {
+    return (this as? AbstractKotlinNativeCompilation)
+        ?.target?.project?.configurations
+        ?.findByName(@Suppress("DEPRECATION") this.defaultSourceSet.implementationMetadataConfigurationName) // KT-58640
+        ?: project.files()
 }

--- a/runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/kotlin/kotlinClasspathUtils.kt
+++ b/runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/kotlin/kotlinClasspathUtils.kt
@@ -6,6 +6,7 @@ import org.jetbrains.dokka.gradle.isAndroidTarget
 import org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet
 import org.jetbrains.kotlin.gradle.plugin.mpp.AbstractKotlinNativeCompilation
 import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompileTool
 
 internal fun Project.classpathOf(sourceSet: KotlinSourceSet): FileCollection {
     val compilations = compilationsOf(sourceSet)
@@ -23,6 +24,26 @@ internal fun Project.classpathOf(sourceSet: KotlinSourceSet): FileCollection {
 }
 
 private fun KotlinCompilation.compileClasspathOf(project: Project): FileCollection {
+    val kgpVersion = project.getKgpVersion()
+
+    // if KGP version < 1.9 or org.jetbrains.dokka.classpath.useOldResolution=true
+    // we will use old (pre 1.9) resolution of classpath
+    if (kgpVersion == null ||
+        kgpVersion < KotlinGradlePluginVersion(1, 9, 0) ||
+        project.classpathProperty("useOldResolution", default = false)
+    ) {
+        return oldCompileClasspathOf(project)
+    }
+
+    return newCompileClasspathOf(project)
+}
+
+private fun KotlinCompilation.newCompileClasspathOf(project: Project): FileCollection {
+    val compilationClasspath = (compileTaskProvider.get() as? KotlinCompileTool)?.libraries ?: project.files()
+    return compilationClasspath + platformDependencyFiles()
+}
+
+private fun KotlinCompilation.oldCompileClasspathOf(project: Project): FileCollection {
     if (this.target.isAndroidTarget()) { // Workaround for https://youtrack.jetbrains.com/issue/KT-33893
         return this.classpathOf(project)
     }

--- a/runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/kotlin/kotlinClasspathUtils.kt
+++ b/runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/kotlin/kotlinClasspathUtils.kt
@@ -12,14 +12,13 @@ internal fun Project.classpathOf(sourceSet: KotlinSourceSet): FileCollection {
     return if (compilations.isNotEmpty()) {
         compilations
             .map { compilation -> compilation.compileClasspathOf(project = this) }
-            .reduce { acc, fileCollection -> acc + fileCollection }
+            .reduce(FileCollection::plus)
     } else {
         // Dokka suppresses source sets that do no have compilations
         // since such configuration is invalid, it reports a warning or an error
         sourceSet.withAllDependentSourceSets()
-            .toList()
             .map { it.kotlin.sourceDirectories }
-            .reduce { acc, fileCollection -> acc + fileCollection }
+            .reduce(FileCollection::plus)
     }
 }
 

